p2p/memp2p/memp2p.go

```diff
@@ -0,0 +1,356 @@
+}
+
+// BroadcastOnChannel sends the message to all peers in the network. It calls
+// parametricBroadcast() with async=false, which means that peers will have
+// their ReceiveMessage() function called synchronously. The call to
+// parametricBroadcast() is done as a go-routine, which means this function is,
+// in fact, non-blocking, but it is identical with BroadcastOnChannelBlocking()
+// in all other regards.
+func (messenger *MemP2PMessenger) BroadcastOnChannel(channel string, topic string, buff []byte) {
+	fmt.Printf("memp2p:BroadcastOnChannel\n")
+	messenger.parametricBroadcast(topic, buff, false)
+}
+
+// Broadcast asynchronously sends the message to all peers in the network. It
+// calls parametricBroadcast() with async=true, which means that peers will
+// have their ReceiveMessage() function independently called as go-routines.
+func (messenger *MemP2PMessenger) Broadcast(topic string, buff []byte) {
+	fmt.Printf("memp2p:Broadcast\n")
+	messenger.parametricBroadcast(topic, buff, true)
+}
+
+// parametricBroadcast sends a message to all peers in the network, with the
+// possibility to choose from asynchronous or synchronous sending.
+func (messenger *MemP2PMessenger) parametricBroadcast(topic string, data []byte, async bool) {
+	if messenger.IsConnectedToNetwork() {
+		message, _ := NewMemP2PMessage(topic, data, messenger.ID())
+		for _, peer := range messenger.Network.Peers() {
+			if async == true {
+				go peer.ReceiveMessage(topic, message)
+			} else {
+				peer.ReceiveMessage(topic, message)
+			}
+		}
+	}
+}
+
+// SendToConnectedPeer sends a message directly to the peer specified by the ID.
+func (messenger *MemP2PMessenger) SendToConnectedPeer(topic string, buff []byte, peerID p2p.PeerID) error {
+	fmt.Printf("memp2p:SendToConnectedPeer\n")
+	if messenger.IsConnectedToNetwork() {
+		if peerID == messenger.ID() {
+			return errors.New("Peer cannot send a direct message to itself")
+		}
+		message, _ := NewMemP2PMessage(topic, buff, messenger.ID())
+		destinationPeer, peerFound := messenger.Network.PeersExceptOne(messenger.ID())[peerID]
+
+		if peerFound == false {
+			return errors.New("Destination peer is not connected to the network")
+		}
+
+		return destinationPeer.ReceiveMessage(topic, message)
+	}
+
+	return errors.New("Peer not connected to network, cannot send anything")
+}
+
+// ReceiveMessage handles the received message by passing it to the message
+// processor of the corresponding topic, given that this Messenger has
+// previously registered a message processor for that topic.
+func (messenger *MemP2PMessenger) ReceiveMessage(topic string, message p2p.MessageP2P) error {
+	fmt.Printf("memp2p:ReceiveMessage\n")
+	messenger.TopicsMutex.Lock()
+	validator, found := messenger.Topics[topic]
+	messenger.TopicsMutex.Unlock()
+
+	if !found {
+		return p2p.ErrNilTopic
+	}
+
+	if validator == nil {
+		return p2p.ErrNilValidator
+	}
+
+	_ = validator.ProcessReceivedMessage(message)
```

Comment:
Should we not check that the returned error is nil?

Reply:
Fixed. Now `ReceiveMessage` returns the error returned by `ProcessReceivedMessage()`.